File: PRACTICAS/practica3/Civitas_P1/src/civitas_p1/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author angel y el javi lee
 */
public class MazoSorpresas {
    
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init(){
        sorpresas = new ArrayList<Sorpresa>();
        cartasEspeciales = new ArrayList<Sorpresa>();
        usadas = 0;
        barajada = false;
        ultimaSorpresa = null;
    }
    
    MazoSorpresas(Boolean debug){
        
        this.debug = debug;
        
        if (debug){
            init();
            Diario.instance.ocurreEvento("Modo debug activado");
        }
        else
            Diario.instance.ocurreEvento("Modo debug no activado");
  
    }
    
    MazoSorpresas(){
        
        init();
        debug = false;
        
    }
    
    int getSize(){
        return sorpresas.size();
    }
    
    void alMazo(Sorpresa sorpresa){
        
        if(!barajada)
            sorpresas.add(sorpresa);
            
    }
    
    Sorpresa siguiente(){
        
        if((!barajada || usadas == sorpresas.size()) && !debug){
            Collections.shuffle(sorpresas);
            usadas = 0;
            barajada = true;
        }
        
        int util = sorpresas.size();
        
        usadas ++;
        
        ultimaSorpresa = sorpresas.get(0);
        
        for(int i = 1; i < util; i++){
            
            sorpresas.set(i - 1, sorpresas.get(i));
            
        }
       
        sorpresas.set(util - 1, ultimaSorpresa);
        
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial (Sorpresa sorpresa){
          
            if (this.sorpresas.contains(sorpresa)){             
                cartasEspeciales.add(sorpresa);
                Diario.instance.ocurreEvento("Se ha inhabilitado la sorpresa " + sorpresa.toString() + " y se ha puesto en cartas especiales.");
                sorpresas.remove(sorpresa);
                
            }
            
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        
        
            if (this.cartasEspeciales.contains(sorpresa)){
                sorpresas.add(sorpresa);
                Diario.instance.ocurreEvento("Se ha habilitado la cartas especial " + sorpresa.toString() + " y se ha puesto en sorpresas.");
                cartasEspeciales.remove(sorpresa);
                
            }
            
    }

    
}

File: PRACTICAS/PRACTICA_ACTUAL/RUBY/pkg/Civitas-0.0.1/lib/mazo_sorpresas.rb
# encoding: utf-8
require_relative 'sorpresa.rb'
require_relative 'diario.rb'

module Civitas
  class MazoSorpresas
    
    attr_reader :ultima_sorpresa
    
    def initialize(d=false)
      @debug=d
      init
      if d
        Diario.instance.ocurre_evento("Modo debug activado")
      else
        Diario.instance.ocurre_evento("Modo debug no activado")
      end
    end

    def init
      @sorpresas=[]
      @cartas_especiales=[]
      @barajada=false
      @usadas=0
      @ultima_sorpresa=nil
    end

    def al_mazo(s)
      @sorpresas.push(s) unless @barajada
    end

    def siguiente
      if !@barajada or @usadas==@sorpresas.size && !@debug
        @usadas=0
        @barajada=true
        @sorpresas.shuffle
      end
      @usadas+=1
      @sorpresas.push(@sorpresas.shift)
      @ultima_sorpresa=@sorpresas[-1]
      @ultima_sorpresa
    end

    def inhabilitar_carta_especial(sorpresa)
      if @sorpresas.include?(sorpresa)
        @sorpresas.delete(sorpresa)
        @cartas_especiales.push(sorpresa)
        Diario.instance.ocurre_evento("Se ha inhabilitado la sorpresa #{sorpresa.to_s} y se ha puesto en cartas especiales.")
      end
    end

    def habilitar_carta_especial(sorpresa)
      if @cartas_especiales.include?(sorpresa)
        @cartas_especiales.delete(sorpresa)
        @sorpresas.push(sorpresa)
        Diario.instance.ocurre_evento("Se ha habilitado la carta especial #{sorpresa.to_s} y se ha puesto en sorpresas.")
      end
    end
  end
end
